modules/electrophysiology_browser/php/sessions.class.inc

```diff
@@ -92,7 +92,11 @@ class Sessions extends \NDB_Page
                 )->withStatus(404);
         }
 
-        if (!$this->_hasAccess($user)) {
+        $candID    = $this->timepoint->getCandID();
+        $candidate = \Candidate::singleton($candID);
+        if (!$this->_hasAccess($user)
+            || !$user->hasProject($candidate->getProjectID())
```

Comment:
@AlexandraLivadas I think it's better to put the project access check inside the hasAccess function

Reply:
@ridz1208 This wouldn't work because the `timepoint` variable is not properly defined when the _hasAccess method is first called. This is also what caused the issue #6557. The _hasAccess method is called first by the Module class [here](https://github.com/aces/Loris/blob/a5304c762454fa8241ce0699af94b2da15f8879d/php/libraries/Module.class.inc#L351), and it is called before the `handle` method is called where the `timepoint` variable is properly defined. So, I couldn't include it in the method because it would give a null pointer error. Let me know if this explanation makes sense!

Reply:
@AlexandraLivadas it makes sense. I think I've seen this usecase before

@HenriRabalais isnt that a similar problem to the one you had with issue tracker? do you remember what Dave suggested toadd to the class in order to instanciate the object before getting to the has access function ?

Reply:
https://github.com/aces/Loris/pull/6471/files

this might be helpful. look at the comment of the `process` function.

Reply:
I tried this approach and it did not work. I think it is because this class (Sessions) is of type `NDB_Page`, while the one you linked is of type `NDB_Form`. If you look at the [Module](https://github.com/aces/Loris/blob/23.0-release/php/libraries/Module.class.inc) class, the method `process` is never called for the actual page. So, when I created the `process` method and defined the timepoint variable in there and then moved the project permission line into the `_hasAccess` method, it gave me a null pointer error again.

Reply:
Shouldn't that mean that line 51 also fails because it accesses timepoint?

Reply:
Yes @johnsaigle, this is addressed in issue #6557 and associated PR #6639. Line 51 checks for site affiliation permissions, which is why in the testing instructions I specify that in order to test this the user needs to have all-site access.

Reply:
OK thanks I didn't see the link previously.

Reply:
I will add it in the PR description just in case!

Reply:
If you need access to a class variable based on request data in the `hasAccess` function, just set it in the `process()` middleware before calling the parent. See modules/instrument_list/php/instrument_list.class.inc for an example.

Reply:
@driusan I sugested that above. @AlexandraLivadas is saying it didnt work

Reply:
All page types should have the `process` middleware called on it, there shouldn't be anything specific to forms in that regard. (Forms are special in that they have a `_process`, but that's different than and unrelated to the PSR15 middleware `process`)

Reply:
@driusan I tried again following a very similar format to `modules/instrument_list/php/instrument_list.class.inc` and it still gives me the following error when I move the project permission to the _hasAccess method:
 ```
PHP Fatal error:  Uncaught Error: Call to a member function getProjectID() on null in /var/www/loris2020/modules/electrophysiology_browser/php/sessions.class.inc:56\nStack trace:\n#0 /var/www/loris2020/php/libraries/Module.class.inc(351): LORIS\\electrophysiology_browser\\Sessions->_hasAccess(Object(User))\n#1 /var/www/loris2020/src/Middleware/ResponseGenerator.php(50): Module->handle(Object(Laminas\\Diactoros\\ServerRequest))\n#2 /var/www/loris2020/src/Middleware/AuthMiddleware.php(63): LORIS\\Middleware\\ResponseGenerator->process(Object(Laminas\\Diactoros\\ServerRequest), Object(LORIS\\electrophysiology_browser\\Module))\n#3 /var/www/loris2020/src/Router/ModuleRouter.php(75): LORIS\\Middleware\\AuthMiddleware->process(Object(Laminas\\Diactoros\\ServerRequest), Object(LORIS\\electrophysiology_browser\\Module))\n#4 /var/www/loris2020/src/Router/BaseRouter.php(105): LORIS\\Router\\ModuleRouter->handle(Object(Laminas\\Diactoros\\ServerRequest))\n#5 /var/www/loris2020/src/Middleware/ResponseGenerator.php(50): LORIS\\Router\\BaseRouter->handle(Object(Laminas in /var/www/loris2020/modules/electrophysiology_browser/php/sessions.class.inc on line 56, referer: https://alivadas-dev.loris.ca/electrophysiology_browser
```

I included the if-statements from the instrument_list.class.inc both in the [_hasAccess](https://github.com/aces/Loris/blob/1f98c1596872adc965d1977f1a2cd23210451363/modules/instrument_list/php/instrument_list.class.inc#L85) and [process](https://github.com/aces/Loris/blob/1f98c1596872adc965d1977f1a2cd23210451363/modules/instrument_list/php/instrument_list.class.inc#L132) methods to check if the timepoint and candidate variables are null. The `\NotFound` error (from _hasAccess) appeared but the `\LorisException` error (from process) did not. So I think that `_hasAccess` is being called even before `process` is called, which is why there is still a null pointer exception.

Reply:
@AlexandraLivadas any news from you investigation ?

Reply:
@h-karim and I have been running into the same problem when trying to implement the `process` method (check out his PR #6639). After some digging, we both found that the `process` method is called **after** the `_hasAccess` method, so this fix would not work.